### PR TITLE
[Snappi] Added multi dut PFC Lossless test case

### PR DESCRIPTION
**Pull request by vikumarks**
### Description of PR
- Testing PFC lossless priorities behavior when sending external pfc pause frames in modular chassis which has multi-asic 

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### What is the motivation for this PR?
To test PFC lossless behavior in multi-dut/multi-asic scenario. 
#### How did you verify/test it?
Tested against voq chassis with SONiC installed in it 
#### Any platform specific information?
Tested against J2C+ 
#### Supported testbed topology if it's a new test case?
T2, T1 & T0 
### Documentation
https://github.com/sonic-net/sonic-mgmt/blob/master/docs/testplan/pfc/PFC_PAUSE_LOSSLESS_README.md

### CODE REVIEW

**Comment by developfast** [comment]
Can you pls add the [Snappi] tag to the title of this PR, and your other Snappi PRs so others can understand the gist of this PR? Here is an example - https://github.com/sonic-net/sonic-mgmt/pull/7718

File: tests/snappi/pfc/test_multidut_pfc_pause_lossless_with_snappi.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+                 tx_port=snappi_ports[1],
+                 tx_port_id=snappi_ports[1]["port_id"],
+                 global_pause=False,
+                 pause_prio_list=pause_prio_list,
+                 test_prio_list=test_prio_list,
+                 bg_prio_list=bg_prio_list,
+                 prio_dscp_map=prio_dscp_map,
+                 test_traffic_pause=True)
+
+    cleanup_config(dut_list, snappi_ports)
```
[the remark this answers is not in the document]
Comment (vikumarks) [reply tests/snappi/pfc/test_multidut_pfc_pause_lossless_with_snappi.py:100]: We  already closed the #7410 , and created  new PR's 7849 for infrastructure changes+ one testcase Global pause.

File: tests/snappi/pfc/test_multidut_pfc_pause_lossless_with_snappi.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+        all_prio_list (pytest fixture): list of all the priorities
+        prio_dscp_map (pytest fixture): priority vs. DSCP map (key = priority).
+        line_card_choice: Line card choice to be mentioned in the variable.py file
+        linecard_configuration_set : Line card classification, (min 1 or max 2  hostnames and asics to be given)
+
+    Returns:
+        N/A
+    """
+    if line_card_choice not in linecard_configuration_set.keys():
+        assert False, "Invalid line_card_choice value passed in parameter"
```
[the remark this answers is not in the document]
Comment (vkuma82) [reply tests/snappi/pfc/test_multidut_pfc_pause_lossless_with_snappi.py]: Instead of assert, using pytest_require

**Comment by kamalsahu0001** [comment]
Test plan for PFC lossless testcase is same for both singledut and multidut. 
https://github.com/sonic-net/sonic-mgmt/blob/master/docs/testplan/pfc/PFC_PAUSE_LOSSLESS_README.md

**Comment by developfast** [comment]
We've recently implemented updates to the Snappi tests, relocating all test cases to the following location: https://github.com/sonic-net/sonic-mgmt/tree/master/tests/snappi_tests. Kindly make the necessary adjustments to reflect this change.

Furthermore, I kindly request you take a look at the pull request at https://github.com/sonic-net/sonic-mgmt/pull/8660. Please ensure that the PFC tests for multi DUT environments align harmoniously with the existing PFC tests.

**Comment by developfast** [comment]
Please add the appropriate new parameters added to this PR to `snappi_test_params`.

**Comment by vikumarks** [comment]
> Please add the appropriate new parameters added to this PR to `snappi_test_params`.
@developfast , Params are part of PR #8186

**Review by vmittal-msft — APPROVED** [review]
Looks ok to me. Can you please add more comments to make code more readable? Also, please share test results and SW version details.

@baiwei0427 @developfast please see if this looks ok.

**Comment by vmittal-msft** [comment]
@baiwei0427 @developfast please see if ok to merge. Looks ok to me.

**Comment by rlhui** [comment]
@vikumarks , are you able to see errors for the test PRs?

**Comment by rlhui** [comment]
these are the errors I saw, please address?
 ==================================== ERRORS ====================================
2024-01-08T22:44:10.9275314Z _ ERROR collecting snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossless_with_snappi.py _
2024-01-08T22:44:10.9275849Z /usr/local/lib/python3.8/dist-packages/_pytest/runner.py:341: in from_call
2024-01-08T22:44:10.9276153Z     result: Optional[TResult] = func()
2024-01-08T22:44:10.9276560Z /usr/local/lib/python3.8/dist-packages/_pytest/runner.py:372: in <lambda>
2024-01-08T22:44:10.9276894Z     call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
2024-01-08T22:44:10.9277383Z /usr/local/lib/python3.8/dist-packages/_pytest/python.py:534: in collect
2024-01-08T22:44:10.9277662Z     return super().collect()
2024-01-08T22:44:10.9278113Z /usr/local/lib/python3.8/dist-packages/_pytest/python.py:455: in collect
2024-01-08T22:44:10.9278437Z     res = ihook.pytest_pycollect_makeitem(
2024-01-08T22:44:10.9278883Z /usr/local/lib/python3.8/dist-packages/pluggy/_hooks.py:493: in __call__
2024-01-08T22:44:10.9279252Z     return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
2024-01-08T22:44:10.9279724Z /usr/local/lib/python3.8/dist-packages/pluggy/_manager.py:115: in _hookexec
2024-01-08T22:44:10.9280052Z     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
2024-01-08T22:44:10.9280583Z /usr/local/lib/python3.8/dist-packages/_pytest/python.py:271: in pytest_pycollect_makeitem
2024-01-08T22:44:10.9280904Z     return list(collector._genfunctions(name, obj))
2024-01-08T22:44:10.9281348Z /usr/local/lib/python3.8/dist-packages/_pytest/python.py:498: in _genfunctions
2024-01-08T22:44:10.9281833Z     self.ihook.pytest_generate_tests.call_extra(methods, dict(metafunc=metafunc))
2024-01-08T22:44:10.9282358Z /usr/local/lib/python3.8/dist-packages/pluggy/_hooks.py:552: in call_extra
2024-01-08T22:44:10.9282704Z     return self._hookexec(self.name, hookimpls, kwargs, firstresult)
2024-01-08T22:44:10.9283168Z /usr/local/lib/python3.8/dist-packages/pluggy/_manager.py:115: in _hookexec
2024-01-08T22:44:10.9283533Z     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
2024-01-08T22:44:10.9284024Z /usr/local/lib/python3.8/dist-packages/_pytest/python.py:150: in pytest_generate_tests
2024-01-08T22:44:10.9284400Z     metafunc.parametrize(*marker.args, **marker.kwargs, _param_mark=marker)
2024-01-08T22:44:10.9284898Z /usr/local/lib/python3.8/dist-packages/_pytest/python.py:1347: in parametrize
2024-01-08T22:44:10.9285188Z     newcallspec = callspec.setmulti(
2024-01-08T22:44:10.9285610Z /usr/local/lib/python3.8/dist-packages/_pytest/python.py:1152: in setmulti
2024-01-08T22:44:10.9285939Z     raise ValueError(f"duplicate {arg!r}")
2024-01-08T22:44:10.9286267Z E   ValueError: duplicate 'reboot_type'

**Comment by developfast** [comment]
@rlhui this PR may have dependency on another unmerged PR or file? @vikumarks please advise if this is the case and the PR #

**Comment by kamalsahu0001** [comment]
Yes. This has dependency on https://github.com/sonic-net/sonic-mgmt/pull/10926<https://urldefense.com/v3/__https:/github.com/sonic-net/sonic-mgmt/pull/10926__;!!I5pVk4LIGAfnvw!ghRKyELMQy3Mb_jxYjpxEedFB13Jxmg9F7FNYnfIesN0ZSqUtOf1avb2YKUfFzW65LNkA4eyizuWsv5XkNOmlstxjSM$>.

From: Dev Ojha ***@***.***>
Date: Tuesday, January 9, 2024 at 6:01 AM
To: sonic-net/sonic-mgmt ***@***.***>
Cc: Kamal Sahu ***@***.***>, Comment ***@***.***>
Subject: Re: [sonic-net/sonic-mgmt] [Snappi] Added multi dut PFC Lossless test case (PR #7912)
@rlhui this PR may have dependency on another unmerged PR or file? @vikumarks please advise if this is the case and the PR # — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented. Message
ZjQcmQRYFpfptBannerStart
This Message is From an External Sender: Use caution opening files, clicking links or responding to requests.

ZjQcmQRYFpfptBannerEnd

@rlhui<https://urldefense.com/v3/__https:/github.com/rlhui__;!!I5pVk4LIGAfnvw!idU3o20Mk3fbh_LAnnkLlak34vY9TB05JtE9-5WFCxx9VeE7-GY9P2vCh3I1Cd9nykri3HEwOwZaKYX6rrQPBlBYNMU$> this PR may have dependency on another unmerged PR or file? @vikumarks<https://urldefense.com/v3/__https:/github.com/vikumarks__;!!I5pVk4LIGAfnvw!idU3o20Mk3fbh_LAnnkLlak34vY9TB05JtE9-5WFCxx9VeE7-GY9P2vCh3I1Cd9nykri3HEwOwZaKYX6rrQPt95d8n0$> please advise if this is the case and the PR #

**Comment by kamalsahu0001** [comment]
@rlhui .. duplicate error issue is fixed. Can you please re-run and see?

**Comment by vmittal-msft** [comment]
@yxieca @wangxin please help merge this well.

**Comment by amitpawar12** [comment]
For multiple line-card scenario, we have following traffic flow direction:
tgen port -------> Ingress LC <-> Egress LC ---------> tgen port

PFC is sent by tgen on the Egress LC
Egress LC <-- PFC -- tgen port

There is verify_pause_frame_count_dut in multidut_helper.py to check the count of the PFC frames on egress line-card. Since the duthost1 and duthost2 are coming from randomized duthost_list in the test-case script, it becomes difficult to identify the egress line-card in multidut_helper.py. We should have some mechanism to identify the right egress line card.

**Comment by amitpawar12** [comment]
The above test does not explicitly set "snappi_extra_params.headroom_test_params" and hence by default gets set to "None".
For lossless scenario, the test_traffic_pause flag is set to "true". The multidut_helper.py has code to check "verify_in_flight_buffer_pkts" if the flag is set to true.

This function is defined in tests/common/snappi_tests/traffic_generation.py and relies on value of snappi_extra_params.headroom_test_params. The code checks if snappi_extra_params.headroom_test_params is set to None and assigns "exceeds_headroom" to False. The above path takes to checking if "tx_bytes_total < dut_buffer_size" and asserts if not true. But tx_bytes_total will totally depend upon the duration of the test. If the test is for longer time or with higher bandwidth, the tx_bytes_total could be greater than dut_buffer_size.
This needs to be addressed.

**Comment by developfast** [comment]
@amitpawar12 this is slightly incorrect. Since we pause the traffic by congesting the egress link out of the DUT even before the first regular frame is TXed from the traffic generator, the switch (in theory) should only be able to hold enough frames as the size of its buffer, since the entire route is 100% congested, and the switch will send back pfc pause frames to the traffic generator to pause. Hence, `tx_bytes_total` here is the number of regular packet bytes sent till the buffer fills up.

**Pull request merged**